### PR TITLE
rt refine: sorries for valid_queues and valid_inQ_queues in Finalise

### PR DESCRIPTION
**Pull request by michaelmcinerney**

### CODE REVIEW

File: proof/refine/ARM/Finalise_R.thy [hunk 1]
```diff
@@ -2955,110 +2955,310 @@ crunch sch_act_not[wp]: cteDeleteOne "sch_act_not t"
...
+  "setQueue tdom prio q \<lbrace>valid_sched_context' sc\<rbrace>"
+  apply (wpsimp simp: setQueue_def valid_sched_context'_def valid_bound_obj'_def
+               split: option.splits)
+  done
+
+lemma threadSet_valid_sched_context'[wp]:
+  "threadSet f y \<lbrace>valid_sched_context' sc\<rbrace>"
+  apply (wpsimp wp: threadSet_wp)
+  by (clarsimp simp: valid_sched_context'_def valid_bound_obj'_def
+                 split: option.splits
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Finalise_R.thy:3075]: Indentation

File: proof/refine/ARM/Finalise_R.thy [hunk 2]
```diff
@@ -3116,21 +3295,153 @@ crunches cancelAllIPC, cancelAllSignals, unbindNotification, unbindMaybeNotifica
...
+   \<lbrace>\<lambda>_. valid_inQ_queues\<rbrace>"
+  unfolding tcbSchedDequeue_def null_def valid_inQ_queues_def
+  apply wp (* stops on threadSet *)
+          apply (rule hoare_post_eq[OF _ threadSet_valid_inQ_queues_dequeue_wp]
+                 , simp add: valid_inQ_queues_def)
+         apply (wp hoare_vcg_if_lift)+
+           apply (wp hoare_vcg_imp_lift setQueue_valid_inQ_queues_except_dequeue_wp
+                     threadGet_const_tcb_at)+
+  (* wp done *)
+  apply (normalise_obj_at')
```
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:3360]: nitpick: no parentheses needed here, right? The same for two lines below

File: proof/refine/ARM/IpcCancel_R.thy [hunk 3]
```diff
@@ -2064,7 +2058,7 @@ lemmas (in delete_one_conc_pre) suspend_makes_simple' =
        suspend_st_tcb_at' [where P=simple', simplified]
 
 crunches cancelSignal, replyUnlink, cleanReply
-  for valid_queus[wp]: valid_queues
+  for valid_queues[wp]: valid_queues
```
Comment (mktnk3) [inline proof/refine/ARM/IpcCancel_R.thy:2061]: thanks for this :)

**Review by mktnk3 — APPROVED** [review]
Nice work! Great to see more sorries fixed!

File: proof/refine/ARM/Finalise_R.thy (hunk 2 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Finalise_R.thy:3360]: Yep, you're right. Removed

**Comment by michaelmcinerney** [comment]
Updated to include a nifty little Eisbach method, and now removes another sorry used in the `delete_one_conc_pre` locale that seems unnecessary. Should be ready to merge now, I think

File: lib/Eisbach_Methods.thy [hunk 4]
```diff
@@ -549,4 +549,7 @@ method solves_emerging methods m1 m2 = (rule bool_protectD, (m1 ; (rule bool_pro
 
 end
 
+method repeat_unless methods stop cont
```
Comment (mktnk3) [inline lib/Eisbach_Methods.thy:552]: This is cool :)

Comment (michaelmcinerney) [reply lib/Eisbach_Methods.thy:552]: Thanks! Though credit to Matt for the actual implementation

Comment (lsf37) [reply lib/Eisbach_Methods.thy:552]: :+1:

**Review by lsf37 — APPROVED** [review]
Looks good!

File: proof/refine/ARM/Finalise_R.thy [hunk 5]
```diff
@@ -3405,10 +3380,8 @@ lemma replyPop_valid_inQ_queues[wp]:
    \<lbrace>\<lambda>_. valid_inQ_queues\<rbrace>"
   (is "valid ?pre _ _")
   apply (clarsimp simp: replyPop_def)
-  apply (rule hoare_seq_ext_skip, wpsimp)
-  apply (rule hoare_seq_ext_skip, wpsimp)
-  apply (rule hoare_seq_ext_skip, wpsimp)
-  apply (rule hoare_seq_ext[OF _ gts_sp'])
+  apply (repeat_unless \<open>rule hoare_seq_ext[OF _ gts_sp']\<close>
+                       \<open>rule hoare_seq_ext_skip, solves wpsimp\<close>)
```
Comment (mbrcknl) [inline proof/refine/ARM/Finalise_R.thy:3384]: Uh huh. :-)

**Review by mbrcknl — APPROVED** [review]
Nice!

File: proof/refine/ARM/Finalise_R.thy [hunk 6]
```diff
@@ -3116,21 +3270,150 @@ crunches cancelAllIPC, cancelAllSignals, unbindNotification, unbindMaybeNotifica
...
+  done
+
+lemma removeFromBitmap_valid_inQ_queues_except[wp]:
+  "removeFromBitmap d p \<lbrace>valid_inQ_queues_except t\<rbrace>"
+  unfolding bitmapQ_defs valid_inQ_queues_except_def
+  by (wpsimp simp: bitmap_fun_defs)+
+
+lemma setQueue_valid_inQ_queues_except_dequeue_wp:
+  "\<And>d p ts t.
+   \<lbrace>\<lambda>s. valid_inQ_queues_except t s \<and> (\<forall>t' \<in> set ts. obj_at' (inQ d p) t' s)
```
Comment (corlewis) [inline proof/refine/ARM/Finalise_R.thy:3312]:
```suggestion
  "\<lbrace>\<lambda>s. valid_inQ_queues_except t s \<and> (\<forall>t' \<in> set ts. obj_at' (inQ d p) t' s)
```

**Review by corlewis — APPROVED** [review]

File: lib/Eisbach_Methods.thy (hunk 4 above)
Comment (mbrcknl) [reply lib/Eisbach_Methods.thy:552]: And credit to @Xaphiosis for the original idea.

File: proof/refine/ARM/Finalise_R.thy (hunk 6 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Finalise_R.thy:3312]: Done!

**Pull request merged**